benchmarks/src/jmh/java/io/grpc/context/ReadWriteBenchmark.java

```diff
@@ -0,0 +1,192 @@
+  @Param({"1", "5", "20"})
+  public int iterations;
+
+  /**
+   * If true we will use the 4-pair batch put, otherwise puts will be individual.
+   */
+  @Param({"true", "false"})
+  public boolean batched;
+
+  /**
+   * Number of possible present keys.
+   */
+  @Param({"4", "10"})
+  public int numKeys;
+
+  /**
+   * The percentage of time we will query for an missing key.
+   */
+  @Param("0.50")
+  public double missingKeyPct;
+
+  /**
+   * Number of keys to put for each iteration.
+   */
+  @Param({"1", "4", "10"})
+  public int writesPerIteration;
+
+  /**
+   * Number of keys to read for each iteration.
+   */
+  @Param({"10"})
+  public int readsPerIteration;
+
+  // Context only supports a 4-pair put function
+  private static final int CONTEXT_MAX_BATCH_SIZE = 4;
+  private static final int DUMMY_VAL = 1;
+  private static final int WORKLOAD_BUFFER_SIZE = 200;
+  // Just an arbitrarily chosen number to avoid always reading the same key.
+  private static final int NUM_MISSING_KEYS = 20;
+
+  private final Random rand = new Random();
+  private Iterator<List<Key<Integer>>> writeWorkload;
+  // Pre-batch the bulk put workload to avoid unnecessary allocation during benchmarks
+  private Iterator<List<List<Key<Integer>>>> writeWorkloadBatched;
+  private Iterator<List<Key<Integer>>> readWorkload;
+
+
+  @Setup
+  public void setUp() throws ExecutionException, InterruptedException, IOException {
+    rand.setSeed(0); // for workload determinism
+    List<Key<Integer>> presentKeys = makeKeys("_p", numKeys);
+    List<Key<Integer>> missingKeys = makeKeys("_m", NUM_MISSING_KEYS);
+
+    // Create a buffer of random read keys and write keys, then partition them up into
+    // the appropriately sized chunks for the test parameters.
+    List<Key<Integer>> writeKeys = new ArrayList<Key<Integer>>();
+    List<Key<Integer>> readKeys = new ArrayList<Key<Integer>>();
+    for (int i = 0; i < WORKLOAD_BUFFER_SIZE; i++) {
+      writeKeys.add(randomPick(presentKeys));
+      readKeys.add(randomPick(rand.nextDouble() < missingKeyPct ? missingKeys : presentKeys));
+    }
+    List<List<Key<Integer>>> writesPerIter = Lists.partition(writeKeys, writesPerIteration);
+    writeWorkload = Iterators.cycle(writesPerIter);
+
+    List<List<List<Key<Integer>>>> writesPerIterbatched = new ArrayList<List<List<Key<Integer>>>>();
+    for (List<Key<Integer>> writeIter : writesPerIter) {
+      writesPerIterbatched.add(Lists.partition(writeIter, CONTEXT_MAX_BATCH_SIZE));
+    }
+    writeWorkloadBatched = Iterators.cycle(writesPerIterbatched);
+
+    readWorkload = Iterators.cycle(Lists.partition(readKeys, readsPerIteration));
+  }
+
+  /**
+   * Missing keys are typically the slow path for Contexts.
+   */
+  @Benchmark
+  @BenchmarkMode(Mode.SampleTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public int readWrite() throws ExecutionException, InterruptedException {
+    int dummy = 0;
+    Context c = Context.ROOT;
+    for (int i = 0; i < iterations; i++) {
```

Comment:
This is suspicious, in that it couples depth of tree with number of different tests (each iteration is actually its own test of sorts, and only the total sum is reported ta the end) with number of gets. I don't think the result of this allows us to make any conclusions.